### PR TITLE
Simplify implementation of TBufferMerger using TFileMerger directly

### PR DESCRIPTION
**Pull request by amadio**
This does not completely resolve the issue of the race conditions, but simplifies the implementation of TBufferMerger a lot by doing away with the ThreadFileMerger imported from GeantV and using TFileMerger directly.

### CODE REVIEW

**Comment by vgvassilev** [comment]
I see that almost nothing is left from the initial version. Shall we change the author attribution, too?

**Comment by amadio** [comment]
Some unrelated commit broke the master (undeclared variable).

**Comment by amadio** [comment]
@vgvassilev While I agree with you about little from the previous implementation being there now, I don't think it's necessary to remove attribution, as the code did come from GeantV. Unless @pcanal and/or Witek want to remove themselves, I prefer to give proper credit for the initial idea.

**Comment by dpiparo** [comment]
Hi, I support being inclusive in recognising credit.

File: io/io/src/TBufferMerger.cxx [hunk 1]
```diff
@@ -60,7 +68,10 @@ void TBufferMerger::WriteOutputFile()
    TDirectoryFile::TContext context;
    TFileMerger merger(false);
 
-   merger.OutputFile(fName, fOption, fCompress);
+   {
+      R__LOCKGUARD2(gROOTMutex);
```
Comment (pcanal) [inline io/io/src/TBufferMerger.cxx]: That one 'should' be spurrious ...

**Comment by amadio** [comment]
This version still unfortunately fails with the stack trace below.
This is true even if I call `mytree->ResetBit(kMustCleanup);` right after creation in the Google test.
```
===========================================================
There was a crash.
This is the entire stack trace of all threads:
===========================================================

Thread 2 (Thread 0x7fbd8b86a700 (LWP 3468)):
#0  0x00007fbd8e6deb7b in __GI___waitpid (pid=3475, stat_loc=stat_loc
entry=0x7fbd8b863080, options=options
entry=0) at ../sysdeps/unix/sysv/linux/waitpid.c:29
#1  0x00007fbd8e6661eb in do_system (line=<optimized out>) at ../sysdeps/posix/system.c:148
#2  0x00007fbd8f7d1e1d in TUnixSystem::Exec (shellcmd=<optimized out>, this=0x111a4f0) at /home/amadio/src/root/core/unix/src/TUnixSystem.cxx:2118
#3  TUnixSystem::StackTrace (this=0x111a4f0) at /home/amadio/src/root/core/unix/src/TUnixSystem.cxx:2412
#4  0x00007fbd8f7d440c in TUnixSystem::DispatchSignals (this=0x111a4f0, sig=kSigSegmentationViolation) at /home/amadio/src/root/core/unix/src/TUnixSystem.cxx:3632
#5  <signal handler called>
#6  TList::Remove (this=this
entry=0x1134310, obj=obj
entry=0x7fbd8b865a90) at /home/amadio/src/root/core/cont/src/TList.cxx:705
#7  0x00007fbd8f7546be in THashList::Remove (this=0x1134310, obj=0x7fbd8b865a90) at /home/amadio/src/root/core/cont/src/THashList.cxx:289
#8  0x00007fbd900270ea in TFileMerger::~TFileMerger (this=0x7fbd8b865a90, __in_chrg=<optimized out>) at /home/amadio/src/root/io/io/src/TFileMerger.cxx:110
#9  0x00007fbd90001798 in ROOT::Experimental::TBufferMerger::WriteOutputFile (this=0x7ffcab4c96f0) at /home/amadio/src/root/io/io/src/TBufferMerger.cxx:63
#10 0x00007fbd8f1d3572 in ?? () from /usr/lib/gcc/x86_64-pc-linux-gnu/6.3.0/libstdc++.so.6
#11 0x00007fbd8e9c43f4 in start_thread (arg=0x7fbd8b86a700) at pthread_create.c:333
#12 0x00007fbd8e70f65f in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:105

Thread 1 (Thread 0x7fbd90beaa80 (LWP 3446)):
#0  0x00007fbd8e9c56cd in pthread_join (threadid=140452066404096, thread_return=0x0) at pthread_join.c:90
#1  0x00007fbd8f1d3427 in std::thread::join() () from /usr/lib/gcc/x86_64-pc-linux-gnu/6.3.0/libstdc++.so.6
#2  0x00007fbd90000dea in ROOT::Experimental::TBufferMerger::~TBufferMerger (this=0x7ffcab4c96f0, __in_chrg=<optimized out>) at /home/amadio/src/root/io/io/src/TBufferMerger.cxx:37
#3  0x0000000000419040 in TBufferMerger_ParallelTreeFill_Test::TestBody (this=<optimized out>) at /home/amadio/src/root/io/io/test/TBufferMerger.cxx:81
#4  0x0000000000440553 in testing::internal::HandleSehExceptionsInMethodIfSupported<testing::Test, void> (location=0x4436da "the test body", method=<optimized out>, object=<optimized out>) at /var/tmp/root/googletest-prefix/src/googletest/googletest/src/gtest.cc:2402
#5  testing::internal::HandleExceptionsInMethodIfSupported<testing::Test, void> (object=object
entry=0x1170670, method=(void (testing::Test::*)(testing::Test * const)) 0x418ec0 <TBufferMerger_ParallelTreeFill_Test::TestBody()>, location=location
entry=0x4436da "the test body") at /var/tmp/root/googletest-prefix/src/googletest/googletest/src/gtest.cc:2438
#6  0x000000000043709a in testing::Test::Run (this=0x1170670) at /var/tmp/root/googletest-prefix/src/googletest/googletest/src/gtest.cc:2475
#7  0x00000000004371e8 in testing::TestInfo::Run (this=0x1171010) at /var/tmp/root/googletest-prefix/src/googletest/googletest/src/gtest.cc:2656
#8  0x00000000004372c5 in testing::TestCase::Run (this=0x1170b90) at /var/tmp/root/googletest-prefix/src/googletest/googletest/src/gtest.cc:2774
#9  0x000000000043759f in testing::internal::UnitTestImpl::RunAllTests (this=0x11708f0) at /var/tmp/root/googletest-prefix/src/googletest/googletest/src/gtest.cc:4649
#10 0x0000000000440a33 in testing::internal::HandleSehExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool> (location=0x442a78 "auxiliary test code (environments or event listeners)", method=<optimized out>, object=<optimized out>) at /var/tmp/root/googletest-prefix/src/googletest/googletest/src/gtest.cc:2402
#11 testing::internal::HandleExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool> (object=0x11708f0, method=(bool (testing::internal::UnitTestImpl::*)(testing::internal::UnitTestImpl * const)) 0x437350 <testing::internal::UnitTestImpl::RunAllTests()>, location=location
entry=0x442a78 "auxiliary test code (environments or event listeners)") at /var/tmp/root/googletest-prefix/src/googletest/googletest/src/gtest.cc:2438
#12 0x00000000004378d0 in testing::UnitTest::Run (this=0x651840 <testing::UnitTest::GetInstance()::instance>) at /var/tmp/root/googletest-prefix/src/googletest/googletest/src/gtest.cc:4257
#13 0x0000000000418789 in RUN_ALL_TESTS () at /var/tmp/root/googletest-prefix/src/googletest/googletest/include/gtest/gtest.h:2233
#14 main (argc=1, argv=0x7ffcab4c9b58) at /var/tmp/root/googletest-prefix/src/googletest/googletest/src/gtest_main.cc:37
===========================================================
```

File: io/io/inc/ROOT/TBufferMerger.hxx [hunk 2]
```diff
@@ -71,17 +76,18 @@ private:
...
+   TBufferMerger &operator=(const TBufferMerger &);
 
    void Push(TBufferFile *buffer);
-   void Listen();
+   void WriteOutputFile();
 
-   std::mutex fFilesMutex;                                       //< Mutex used to lock fAttachedFiles
+   const std::string fName;
+   const std::string fOption;
+   const Int_t fCompress;
```
Comment (vgvassilev) [inline io/io/inc/ROOT/TBufferMerger.hxx]: Is there a particular reason to mark the members only `const`?

Comment (amadio) [inline io/io/inc/ROOT/TBufferMerger.hxx]: As opposed to what?

Comment (vgvassilev) [inline io/io/inc/ROOT/TBufferMerger.hxx]: As opposed to not being `const`. What's the meaning of const int/string? The compiler can deduce this based on the fact that those are allocated on the stack so if it is there for optimization I think it is redundant. If you are telling the reader that you only touch this in the constructors, could be good to know. I don't think this is a common practice in ROOT, however.

Comment (pcanal) [inline io/io/inc/ROOT/TBufferMerger.hxx]: > Is there a particular reason to mark the members only const?
It insures that the invariant of the class (not-allowed to change output file after the class) are guaranteed.  It also allows/indicates-that-you-are-allowed to use those from multiple thread without risk of race condition.

Comment (amadio) [inline io/io/inc/ROOT/TBufferMerger.hxx]: Yes, what @pcanal said, and it makes it easy to catch bugs if these members are ever modified by mistake.

**Comment by amadio** [comment]
The Mac errors are not related to this PR.

File: io/io/inc/ROOT/TBufferMerger.hxx (hunk 2 above)
Comment (vgvassilev) [inline io/io/inc/ROOT/TBufferMerger.hxx]: Ok, make sense. I see their semantics is readonly.

**Pull request merged**